url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/canvas/mxVmlCanvas.java

```diff
@@ -0,0 +1,604 @@
+ * Copyright (c) 2007, Gaudenz Alder
+ */
+package com.mxgraph.canvas;
+
+import java.awt.Rectangle;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Map;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+
+import com.mxgraph.util.mxConstants;
+import com.mxgraph.util.mxPoint;
+import com.mxgraph.util.mxRectangle;
+import com.mxgraph.util.mxUtils;
+import com.mxgraph.view.mxCellState;
+
+/**
+ * An implementation of a canvas that uses VML for painting.
+ */
+public class mxVmlCanvas extends mxBasicCanvas
+{
+
+	/**
+	 * Holds the HTML document that represents the canvas.
+	 */
+	protected Document document;
+
+	/**
+	 * Constructs a new VML canvas for the specified dimension and scale.
+	 */
+	public mxVmlCanvas()
+	{
+		this(null);
+	}
+
+	/**
+	 * Constructs a new VML canvas for the specified bounds, scale and
+	 * background color.
+	 */
+	public mxVmlCanvas(Document document)
+	{
+		setDocument(document);
+	}
+
+	/**
+	 * 
+	 */
+	public void setDocument(Document document)
+	{
+		this.document = document;
+	}
+
+	/**
+	 * Returns a reference to the document that represents the canvas.
+	 * 
+	 * @return Returns the document.
+	 */
+	public Document getDocument()
+	{
+		return document;
+	}
+
+	/**
+	 * 
+	 */
+	public void appendVmlElement(Element node)
+	{
+		if (document != null)
+		{
+			Node body = document.getDocumentElement().getFirstChild()
+					.getNextSibling();
+
+			if (body != null)
+			{
+				body.appendChild(node);
+			}
+		}
+
+	}
+
+	/* (non-Javadoc)
+	 * @see com.mxgraph.canvas.mxICanvas#drawCell()
+	 */
+	public Object drawCell(mxCellState state)
+	{
+		Map<String, Object> style = state.getStyle();
+		Element elem = null;
+
+		if (state.getAbsolutePointCount() > 1)
+		{
+			List<mxPoint> pts = state.getAbsolutePoints();
+
+			// Transpose all points by cloning into a new array
+			pts = mxUtils.translatePoints(pts, translate.x, translate.y);
+
+			// Draws the line
+			elem = drawLine(pts, style);
+			Element strokeNode = document.createElement("v:stroke");
+
+			// Draws the markers
+			String start = mxUtils.getString(style,
+					mxConstants.STYLE_STARTARROW);
+			String end = mxUtils.getString(style, mxConstants.STYLE_ENDARROW);
+
+			if (start != null || end != null)
+			{
+				if (start != null)
+				{
+					strokeNode.setAttribute("startarrow", start);
+
+					String startWidth = "medium";
+					String startLength = "medium";
+					double startSize = mxUtils.getFloat(style,
+							mxConstants.STYLE_STARTSIZE,
+							mxConstants.DEFAULT_MARKERSIZE)
+							* scale;
+
+					if (startSize < 6)
+					{
+						startWidth = "narrow";
+						startLength = "short";
+					}
+					else if (startSize > 10)
+					{
+						startWidth = "wide";
+						startLength = "long";
+					}
+
+					strokeNode.setAttribute("startarrowwidth", startWidth);
+					strokeNode.setAttribute("startarrowlength", startLength);
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 112:136, 138:162.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.